bindings/c/test/mako/utils.c

```diff
@@ -67,15 +68,17 @@ int digits(int num) {
 }
 
 /* generate a key for a given key number */
+/* prefix is "mako" by default, prefixpadding = 1 means 'x' will be in front rather than trailing the keyname */
 /* len is the buffer size, key length + null */
-void genkey(char* str, int num, int rows, int len) {
-	int i;
-	int rowdigit = digits(rows);
-	sprintf(str, KEYPREFIX "%0.*d", rowdigit, num);
-	for (i = (KEYPREFIXLEN + rowdigit); i < len - 1; i++) {
-		str[i] = 'x';
-	}
-	str[len - 1] = '\0';
+void genkey(char* str, char* prefix, int prefixlen, int prefixpadding, int num, int rows, int len) {
+        const int rowdigit = digits(rows);
+        const int prefixoffset = prefixpadding ? len - (prefixlen + rowdigit) - 1 : 0;
+        char* prefixstr = (char*)malloc(sizeof(char) * (prefixlen + rowdigit + 1));
```

Comment:
It turns out that `snprintf()` doesn't work this way, and would skip the final digit of num.  So we can fix that by doing `str[prefixOffset + prefixlen + rowdigit - 1] = num % 10;`

Reply:
New idea that is less ugly and just avoids the heap allocation but still copies.

Just change `malloc` to `alloca` and remove the free on line 81.  This will use stack space for the temporary string so it's freed upon return.